File: frontend/traceability_tutor/src/app/models/project.ts
import {IterationDTO, LevelDTO, ProjectDTO} from "../../../gen/model";

export class Project implements Omit<ProjectDTO, 'levels'> {
    id: number
    name: string;
    owner: number;
    repoUrl: string;
    repoName: string;
    iterations: IterationDTO[] = [];
    levels: Map<string, LevelDTO> = new Map();
    lastOpened: Date;

    constructor(projectDTO: ProjectDTO) {
        this.id = projectDTO?.id;
        this.name = projectDTO.name;
        this.owner = projectDTO.owner;
        this.repoUrl = projectDTO?.repoUrl || '';
        this.repoName = projectDTO?.repoName || '';
        projectDTO?.levels.forEach(level => {
            let levelName = level.name.toLowerCase();
            levelName = levelName[0].toUpperCase() + levelName.substring(1);
            this.levels.set(levelName, level);
        });
        this.lastOpened = projectDTO.lastOpened;
    }

    addIteration(iterationDTO: IterationDTO) {
        this.iterations.push(iterationDTO);
    }

    toDto(): ProjectDTO {
        return {
            ...this,
            levels: Array.from(this.levels.values())
        }
    }

}

File: src/main/java/uniba/fmph/traceability_tutor/domain/Project.java
package uniba.fmph.traceability_tutor.domain;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.validator.constraints.URL;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;


@Entity
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@Builder
@AllArgsConstructor
public class Project {

    @Id
    @Column(nullable = false, updatable = false)
    @SequenceGenerator(
            name = "primary_sequence",
            sequenceName = "primary_sequence",
            allocationSize = 1,
            initialValue = 10000
    )
    @GeneratedValue(
            strategy = GenerationType.SEQUENCE,
            generator = "primary_sequence"
    )
    private Long id;

    @Column(nullable = false)
    @URL
    private String repoUrl;

    @Column(nullable = false)
    private String repoName;

    @Column(nullable = false)
    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Item> projectItems;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Relationship> projectRelationships;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Iteration> iterations;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private OffsetDateTime dateCreated;

    @LastModifiedDate
    @Column(nullable = false)
    private OffsetDateTime lastUpdated;

    @Column
    private OffsetDateTime lastOpened;

    @OneToOne(mappedBy = "project", cascade = {CascadeType.MERGE, CascadeType.PERSIST}, orphanRemoval = true, fetch = FetchType.EAGER)
    private UserSecret userSecret;

    @Column
    private OffsetDateTime lastCodeFetched;

    public Project() {

    }

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    private List<Level> levels = new ArrayList<>();

    public void setLevels(List<Level> levels) {
        if (this.levels == null) this.levels = new ArrayList<>();
        if (levels != null) {
            levels.forEach(this::addLevel);
        }
    }

    public void addLevel(Level level) {
        levels.add(level);
        level.setProject(this);
    }

    public void removeLevel(Level level) {
        levels.remove(level);
        level.setProject(null);
    }

}
